Add formatting of `Cargo.toml` files as per convention

thomaseizinger opened the pull request:
Draft PR to showcase what I currently have.

Will eventually resolve https://github.com/thomaseizinger/dprint-plugin-cargo-toml/issues/1.

A few notes:

1. I opted to add the rustfmt plugin to this repo. Let me know if you don't want that as part of this PR.
2. I had to fork `taplo` to bump `rowan` to the latest pre-release. This is necessary to get access to the new "mutation" API of rowan which makes life a lot easier for mutating the syntax tree.
3. The formatting works for basic usecases. The testcase currently only fails because the trialing comment gets eaten somehow. If you have a spare minute, do mind having a look at that @dsherret?

thomaseizinger commented on tests/specs/CargoToml/CargoToml_All.txt:
```diff
@@ -0,0 +1,108 @@
+serdebug = "1"
+strum = "0.19"
+strum_macros = "0.19"
+testcontainers = { version = "0.10", optional = true } # In dependencies instead of dev-dependencies so it can be set as optional, only used for test purposes
```
The test fails because this comment is not present in the output. I don't understand why. When I `dbg!` the nodes in `Section::insert`, it it still there :man_shrugging: 

cc @dsherret

thomaseizinger commented on src/format_text.rs:
```diff
@@ -3,16 +3,21 @@ use super::parser::parse_items;
+pub fn format_text(file_path: &Path, text: &str, config: &Configuration) -> Result<String, ErrBox> {
     let node = parse_taplo(text)?;
 
+    let node = if file_path.ends_with("Cargo.toml") {
```
Should we guard this behind a config option as well?

dsherret replied:
Yeah, maybe let's just call it `cargo` and have it `true` by default? We could do it in a separate PR too. This one seems good to merge now if you think it's ready.

thomaseizinger replied:
I am having 2nd thoughts on this, I don't think it needs to be configurable. I don't see why one would opt out of this ever. The formatting is pretty non-invasive and follows official recommendations.

After all, opt-out is already possible by simply excluding `Cargo.toml`s explicitly.

dsherret replied:
Sure. Yeah, we can leave it out until someone asks for it.

thomaseizinger commented on src/format_text.rs:
```diff
@@ -53,3 +137,26 @@ fn config_to_print_options(text: &str, config: &Configuration) -> PrintOptions {
+
+impl SyntaxNodeExt for SyntaxNode {
+    fn entry_key_text(&self) -> String {
+        let key = self
+            .children()
+            .find(|child| child.kind() == SyntaxKind::KEY)
+            .expect("ENTRY should contain KEY");
+
+        let ident = key
+            .children_with_tokens()
+            .find_map(|child| match child {
+                SyntaxElement::Token(token) if token.kind() == SyntaxKind::IDENT => Some(token),
+                _ => None,
+            })
+            .expect("KEY should contain IDENT");
```
Are these panics ok?

dsherret replied:
Yup!

thomaseizinger commented on src/format_text.rs:
```diff
@@ -45,6 +50,85 @@ fn parse_taplo(text: &str) -> Result<SyntaxNode, String> {
+}
+
+fn sort_cargo_package_section(left: &SyntaxNode, right: &SyntaxNode) -> Ordering {
+    match (
+        left.entry_key_text().as_str(),
+        right.entry_key_text().as_str(),
+    ) {
+        ("name", _) => Ordering::Less,
+        ("version", "name") => Ordering::Greater,
+        ("version", _) => Ordering::Less,
+        ("description", _) => Ordering::Greater,
+        (_, "name") => Ordering::Greater,
+        (_, "version") => Ordering::Greater,
+
+        (left, right) => left.cmp(right),
+    }
```
Any input on making this easier to understand is highly appreciated.

The requirements are:

- `name` comes first
- `version` is second
- `description` comes last
- everything in between alphabetically sorted.

dsherret commented:
@thomaseizinger thanks for this! How difficult would it be to remove the formatting changes in the PR as it adds a lot of noise? Then we could do a separate PR for that.

thomaseizinger commented:
> @thomaseizinger thanks for this! How difficult would it be to remove the formatting changes in the PR as it adds a lot of noise? Then we could do a separate PR for that.

Not difficult at all, just need to drop one commit :)
The commits are atomic though, so you can review them independently and just skip the formatting one!

Let me know if that would also work for you.

thomaseizinger commented:
> @thomaseizinger thanks for this! How difficult would it be to remove the formatting changes in the PR as it adds a lot of noise? Then we could do a separate PR for that.

Done here: https://github.com/dprint/dprint-plugin-toml/pull/2

dsherret commented:
LGTM! Thanks so much @thomaseizinger. Let me know if you think this is ready now (looks ready to me... I did a few small fixes and fixed that comment issue here https://github.com/dprint/dprint-plugin-toml/commit/53d8c336db82c45c3712a626eb19e2768d2c817b

thomaseizinger commented on Cargo.toml:
```diff
@@ -26,9 +26,10 @@ tracing = ["dprint-core/tracing"]
 [dependencies]
 dprint-core = { version = "0.43.0", features = ["formatting"] }
-taplo = { version = "0.6.1", default_features = false }
+taplo = { git = "https://github.com/thomaseizinger/taplo", default_features = false }
```
Should we (git-subtree) vendor this into the workspace for now to make the necessary changes? Or do you want to fork it into the dprint org?

thomaseizinger replied:
@dsherret I think we need a least a specific revision here or track Cargo.lock!

But I think I'd prefer vendoring it into the workspace. Would you be okay with that?

dsherret replied:
Oh, good question... we need it to work to publish to cargo.

> This is necessary to get access to the new "mutation" API of rowan which makes life a lot easier for mutating the syntax tree.

How hard is it to do this the harder way? I'd rather not maintain and publish a fork of taplo. Maybe we can ask the rust-analyzer people when they're planning to release 0.13.0.

dsherret replied:
> But I think I'd prefer vendoring it into the workspace. Would you be okay with that?

Do you mean just including all the source code from taplo in a folder under `src/taplo` then using that as a module? That works as a temporary measure long as you include their license and copyright. Maybe lets do that in a separate PR if that's what you mean?

thomaseizinger replied:
> Maybe we can ask the rust-analyzer people when they're planning to release 0.13.0.

Inquired about that here: https://github.com/rust-analyzer/rowan/issues/110
Request for depending on the RC in taplo is here: https://github.com/tamasfe/taplo/issues/147

> Oh, good question... we need it to work to publish to cargo.

Do we need to publish to crates.io?

For building a wasm binary and uploading it to GitHub releases, we don't need the crate to be on crates.io.

dsherret replied:
I don't think it makes sense for taplo to bother depending on a release candidate when it doesn't need it. Thanks for opening the rowan issue!

Yeah, I've been publishing these plugins on crates.io so they can be used by other Rust applications. For example, Deno depends on dprint-plugin-typescript, markdown, and json. Perhaps toml formatting could be added in the future.

thomaseizinger commented:
@dsherret I've rebased onto latest main and squashed your changes into the original commit.

Additionally, I've fixed the revision of the fork to a concrete one, just to avoid unintended changes.

If you are happy with merging this, go for it. I am not yet sure how we should resolve the dependency situation.

dsherret commented:
Thanks again @thomaseizinger! We can deal with the dependency issue in the future. I'll try to do a non-cargo release tomorrow.
